Update specs for Signature validator

akshay-ap opened the pull request:
Fixes #43 

This PR gives details of validation of EIP712 signatures only.

Changes in PR:
- Create Sequence diagram for Signature Validator flow
- Add a new module type for Signature Validator and  Signature Validator Hooks
- Introduce hooks interface for signature validators see `ISignatureValidatorHooks`
- Change interface ISafeProtocol712SignatureValidator
   `isValidSignature(...)` function has `signatures` parameters
- Add new interface `ISafeProtocolSignatureValidatorManager`
     - Define layout of the `data` field in `handle(...)` function of signature validator as function handler
- Describe default validation flow and domain specific validation flow
- Describe how to enable a signature validator, setting signature validator as function handler in SafeProtocolManager 
- Describe expected encoding for default validation flow and domain specific validation flow
- Add note on security consideration when a malicious actor changes routing

rmeissner commented on modules/README.md:
```diff
@@ -146,4 +149,53 @@ interface ISafeProtocol712SignatureValidator {
+     * @param dataHash Hash of the data that is signed
+     * @param data Arbitrary data containing the following layout:
+     *             Layout of the data:
+     *              0x00 - 0x04: selector
```
Do we have any more information on this layout? 

From a first glance I would also argue that it is missing the signature

akshay-ap replied:
updated interface

rmeissner commented on modules/README.md:
```diff
@@ -146,4 +149,53 @@ interface ISafeProtocol712SignatureValidator {
+	participant RegistryContract
+	participant SignatureValidatorContract
+
+	Account->>SignatureValidatorManager: Set SignatureValidatorContract
```
I would move the setup part to a separate diagram

rmeissner commented on modules/README.md:
```diff
@@ -146,4 +149,53 @@ interface ISafeProtocol712SignatureValidator {
+	participant SignatureValidatorContract
+
+	Account->>SignatureValidatorManager: Set SignatureValidatorContract
+		SignatureValidatorManager->>RegistryContract: Check if contract is listed and not flagged
```
contract is the signature validator here, right?

akshay-ap replied:
Here its the `SignatureValidatorManager `. 
`SignatureValidatorManager` will check if there exists a `SignatureValidator` for given domain, if yes, it will further check in `Registry` if `SignatureValidator` is listed and not flagged.

I will update the specs

rmeissner reviewed (changes requested):
Following use cases are not covered for me (will document this also in the attached issue):
- How to handle multiple signature validators
- How does this relate to 4337 (as there is also signature validation necessary)

akshay-ap commented:
> Following use cases are not covered for me (will document this also in the attached issue):
> 
> * How to handle multiple signature validators
> * How does this relate to 4337 (as there is also signature validation necessary)

Added details on handling multiple signatures validators
Regarding 4337, it hasn't been considered so far yet. We have a followup task to be covered in next iterations.

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+     * @param sender Address requesting signature validation
+     * @param data Calldata containing the function selector, signature hash, domain separator, type hash, encoded data and payload forwarded by the Manager
+     *              Arbitrary data containing the following layout:
+     *              Layout of the data:
```
Can we please document the structure of this data somewhere outside of the interface comment in a easy to read way

akshay-ap replied:
added table

akshay-ap replied:
replaced with solidity code

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+     *                  0x68 to 0x9a: typeHash
+     *                  0x9a to 0xcc: encodeData length
+     *                  0xcc to <0xcc + encodeData length>: encodeData
+     *                  <0xcc + encodeData length> to <0xcc + encodeData length> + 0x32 : signature length
```
shouldn't this be `<0xcc + encodeData length + 0x20>`?

In most of these fields it seems that a decimal used even so it is a hex representation. A hash is 32 bytes wide, so 0x20 in hex encoding.

akshay-ap replied:
ISafeProtocol712SignatureValidator

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
         bytes calldata encodeData,
-        bytes calldata payload
+        bytes calldata signature
+        bytes calldata additionalData
```
Do we have an example for possible additional Data?

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
         bytes32 domainSeparator,
         bytes32 typeHash,
```
What is the advantage of having `typeHash` and `encodedData` via just a `messageHash`?

rmeissner replied:
Also can a `ISafeProtocol712SignatureValidator` assume that the caller has checked that `_hash` is the 712 hash of the other params?

akshay-ap replied:
> What is the advantage of having `typeHash` and `encodedData` via just a `messageHash`?

A signature validator as per current spec is specific only to a domain. It should ideally support any type of signed message. If typehash is hardcoded in the validator contract then, hash validation would be specific to domain and type. I think it would be too restrictive in that case.

So, by expecting a `typeHash` and encoded message as a validation parameter, signature validator can validate hash for any typed data

akshay-ap replied:
> Also can a `ISafeProtocol712SignatureValidator` assume that the caller has checked that `_hash` is the 712 hash of the other params?

I propose that we do it in Signature validator manager. Added in spec (sequence diagram).

akshay-ap replied:
removed typeHash and encodeData

rmeissner commented on modules/README.md:
````diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+
+```solidity
+interface ISignatureValidatorHook {
+    function preValidationHook(address account, bytes32 domain, bytes payload) returns (bool success, bytes32 result) external;
````
What is payload?

rmeissner replied:
wouldn't it make sense to also mention which validator is used?

rmeissner commented on modules/README.md:
````diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+    SignatureValidatorManager-->>Account: Ok
+```
+
+### Signature validator hooks
````
Should this be part of the hooks section?

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+     *                  <0xcc + encodeData length> + 0x32 + signature length to <0xcc + encodeData length> + 0x32 + signature length + 0x32 : additional data length
+     *                  <0xcc + encodeData length> + 0x32 + signature length + 0x32 to end: additional data
+     */
+    function handle(
```
Why do we need both the `handle` and `isValidSignature` function?

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+	ExternalContract->>Account: Check if signature is valid for the account (Call isValidSignature(bytes32,bytes))
+    Account->>SignatureValidatorManager: Call isSignatureValid(bytes32,bytes)
+    SignatureValidatorManager->>SignatureValidatorManager: Decode data and extract domain, typeHash, encodeData and payload
+    SignatureValidatorManager->>SignatureValidatorManager: Check if Signature Validator contract is enabled for the domain
```
Isn't this more a "load Signature Validator for domain"?

rmeissner replied:
Also what happens if no signature validator is enabled?

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+
+	ExternalContract->>Account: Check if signature is valid for the account (Call isValidSignature(bytes32,bytes))
+    Account->>SignatureValidatorManager: Call isSignatureValid(bytes32,bytes)
+    SignatureValidatorManager->>SignatureValidatorManager: Decode data and extract domain, typeHash, encodeData and payload
```
What if it is not an eip712 based signature?

akshay-ap replied:
revert

rmeissner commented on modules/README.md:
```diff
@@ -116,8 +117,10 @@ Kudos to @mfw78
+
 References:
-- https://github.com/zerodevapp/kernel/blob/main/src/validator/IValidator.sol
+- https://github.com/zerodevapp/kernel/blob/main/src/interfaces/IValidator.sol
```
Should we move this to the end of the signature validator section? This was originally added as a something to get inspiration from. Also might be worth adding a reference to 6900.

rmeissner commented on modules/README.md:
```diff
@@ -141,9 +145,123 @@ interface ISafeProtocol712SignatureValidator {
+     * @param data Calldata containing the function selector, signature hash, domain separator, type hash, encoded data and payload forwarded by the Manager
+     *              Arbitrary data containing the following layout:
+     *              Layout of the data:
+     *                  0x00 to 0x04: 4 bytes function selector
```
What function selector is this?

akshay-ap replied:
This is the 4 bytes of function selector for which the Signature validator manager is enabled as a function handler in the Protocol Manager. e.g. `isValidSignature(bytes32,bytes)`

rmeissner commented on modules/README.md:
```diff
@@ -116,8 +117,10 @@ Kudos to @mfw78
 
 There are continuous efforts to expand the types of signatures supported by the EVM beyond the currently predominant secp256k1 elliptic curve. For example, a signature scheme gaining popularity is based on the secp256r1 elliptic curve (see EIP-7212). Signature Validators allow accounts to support new standards and enable use-cases such as Passkeys-enabled smart accounts, BLS/Schnorr or quantum-se ...
 
+When a contract wants to verify account signature, it should call `isValidSignature(bytes32,bytes)` on the account and the account calls the `SignatureValidatorManager` for validation.
```
Should we reference ERC-1271 here?

rmeissner commented on modules/README.md:
```diff
@@ -116,8 +117,10 @@ Kudos to @mfw78
+- https://github.com/zerodevapp/kernel/blob/main/src/interfaces/IValidator.sol
 
 
 - EIP-712 based Signature Validator
```
As we don't invision another signature validator right now, I would remove the mention of ERC-712 in the interface names. I would only keep them for the explanation of the parameters.

rmeissner reviewed (changes requested):
There are still a lot of open questions.

I think it would also be helpful to create a diagram that put this manager into context with the existing manager. As mentioned before for me this should all happen within the same contract, but would be nice to document considerations for this (might be more on the implementation side).

I would also propose to restructure this spec section. To start with after the general motivation that is already there I would start with explaining the general flow supported by the validation sequence diagram.

After that a short recap of the most important parts of ERC-712 (domain) and the mapping based on the domain could be explained. Here also the encoding of the signature should be outlined, as this is essential for the validation flow.

Then maybe have a section on the validator hooks and outline how these can improve security and why this is important.

Once all the components have been explained it can be explained how these are combined in the manager.

Only after that I would have a section with the Solidity interfaces.

akshay-ap commented:
> There are still a lot of open questions.
> 
> I think it would also be helpful to create a diagram that put this manager into context with the existing manager. As mentioned before for me this should all happen within the same contract, but would be nice to document considerations for this (might be more on the implementation side).
> 
> I would also propose to restructure this spec section. To start with after the general motivation that is already there I would start with explaining the general flow supported by the validation sequence diagram.
> 
> After that a short recap of the most important parts of ERC-712 (domain) and the mapping based on the domain could be explained. Here also the encoding of the signature should be outlined, as this is essential for the validation flow.
> 
> Then maybe have a section on the validator hooks and outline how these can improve security and why this is important.
> 
> Once all the components have been explained it can be explained how these are combined in the manager.
> 
> Only after that I would have a section with the Solidity interfaces.

This [commit](https://github.com/safe-global/safe-core-protocol-specs/pull/44/commits/30f4e53883331aa50be60c3b8f2cb54244b9ec2e) addresses most of the points mentioned

rmeissner reviewed (changes requested):
I really like the context that was added. 

I posted a proposal to add more context for the signature encoding and routing.

akshay-ap commented on modules/README.md:
```diff
@@ -115,35 +139,175 @@ Kudos to @mfw78
+
+### SignatureValidatorManager with SafeProtocolManager
+
+An account can either implement logic to call `SignatureValidatorManager` for validating signature or set `SafeProtocolManager` as a fallback handler and set `SignatureValidatorManager` as function handler for [isValidSignature(bytes32,bytes)](https://eips.ethereum.org/EIPS/eip-1271) function in the manager. So, when a contract wants to verify account signature, it should call `isValidSignature(b ...
```
Here, an alternative approach to using function handler is mentioned for accounts that are not Safe account or something that we can consider having in Safe v2

akshay-ap replied:
removed and spec now mention only 1 way i.e., using function handler

akshay-ap commented on modules/README.md:
```diff
@@ -115,35 +139,175 @@ Kudos to @mfw78
+- If signature validator hook is enable, call to `postValidationHook(...)` reverted
+- Transaction ran out of gas
+
+The Layout of the encoded data received by the signature validator is expected to be as follows:
```
Should we also use `IValidator` and rename interface function to `validate(...)`? or its just for reference?

akshay-ap replied:
Thanks! Add context in this commit: https://github.com/safe-global/safe-core-protocol-specs/pull/44/commits/95ded768be98581debfcc4691e0c1b0e99b107cf

akshay-ap replied:
> Note to Safe Protocol Team: we should evaluate the risk of the malicious removal of the bytes4 identifier. This allows an attacker to choose if a signature is validated by the default mechanism or via a signature validator.

Indeed we need 4 bytes identifier along with a default validator for validating messages signed by a safe. Updated specs accordingly

rmeissner replied:
> Should we also use IValidator and rename interface function to validate(...)? or its just for reference?

Was just for reference, we should align the examples to be consistent with our spec

akshay-ap commented on modules/README.md:
```diff
@@ -115,35 +139,175 @@ Kudos to @mfw78
      * @param safe The Safe that has delegated the signature verification
      * @param sender The address that originally called the Safe's `isValidSignature` method
-     * @param _hash The EIP-712 hash whose signature will be verified
+     * @param hash The EIP-712 hash whose signature will be verified
```
using `messageHash`

akshay-ap commented on modules/README.md:
```diff
@@ -76,6 +79,27 @@ interface ISafeProtocolHooks {
+    /**
+     * @param account Address of the account for which signature is being validated
+     * @param validator Address of the validator contract to be used for signature validation
+     * @param data Bytes containing domain, typeHash, encodeData length, encodeData, signature length, signature.
```
We cannot have distinct params after introducing default validator. As the signed message may be non 712 messages.

rmeissner commented on modules/README.md:
```diff
@@ -244,7 +264,7 @@ interface ISafeProtocolSignatureValidator {
      * @param safe The Safe that has delegated the signature verification
      * @param sender The address that originally called the Safe's `isValidSignature` method
-     * @param hash The EIP-712 hash whose signature will be verified
+     * @param messageHash The EIP-712 hash whose signature will be verified
```
`messageHash` is now used for different purposes in this PR.

In the example on the routing `messageHash` refers to the result of `hashStruct` of the EIP-712 spec (we could rename this to `structHash`).

And here it refers to the final EIP-712 hash.

akshay-ap replied:
done

rmeissner commented on modules/README.md:
```diff
@@ -202,18 +194,46 @@ The possible cases for a transaction to revert are:
-| 0x84                                                        | <0x84 + encodeData length>                                  | encodeData                |
-| <0x84 + encodeData length>                                  | <0x84 + encodeData length> + 0x20                           | signature length          |
-| <0x84 + encodeData length> + 0x20                           | <0x84 + encodeData length> + 0x20 + signature length        | signature                 |
+    | 0x00                                                        | 0x04                                                        | 4 bytes selector |
```
In the example above we propose `abi.encode`, here we use some custom encoding. Which one is it?

akshay-ap replied:
removed the table with solidity code showing how encoding should be done

rmeissner commented on modules/README.md:
````diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+The encoded data to be received by the signature validator manager as `bytes` parameter of `isValidSignature(bytes32,bytes)` is expected to be encoded as follows:
+
+```solidity
+function encodeData(bytes4 selector, bytes32 domain, bytes32 hashStruct, bytes calldata signatures) public pure returns (bytes memory) {
````
what is the selector? 

Also isn't this the same as `abi.encodeWithSelector`? 

Could we use `abi.encodeWithSignature` and define it like a method call?

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+
+### SignatureValidatorManager with SafeProtocolManager
+
+An account must set `SafeProtocolManager` as a fallback handler and set `SignatureValidatorManager` as function handler for [isValidSignature(bytes32,bytes)](https://eips.ethereum.org/EIPS/eip-1271) function in the manager. So, when a contract wants to verify account signature, it should call `isValidSignature(bytes32,bytes)` on the account and the `SafeProtocolManager` will forward the call to t ...
```
This adds quite some gas overhead, is there a way to optimize this?

rmeissner replied:
Potentially it makes sense to not mention anything about `ISafeProtocolSignatureValidatorManager.handle` and how this is set. For the protocol in itself this is not really important. How the validation  is routed and the interface of the validator or the only necessary facts for the spec, the other thing is a implementation detail.

That being said, I think we need to iterate on what is part of the Manager and what is delegated to other contracts.

akshay-ap replied:
If we remove these details, the whole section of `SignatureValidatorManager with SafeProtocolManager` will be removed and it is not clear how `isValidSignature(bytes32,bytes)` is connected to validation flow

rmeissner commented on modules/README.md:
````diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+}
+```
+
+To distinguish signatures that follow this format from "default" signatures the signature is prepended with a special bytes4 identifier. When the dapp requests the signature the wallet can then create the signature in the required format for the correct signature routing.
````
Should we mention that the "default" handling depends on the account implementation?

akshay-ap replied:
added in this commit https://github.com/safe-global/safe-core-protocol-specs/pull/44/commits/eabb02f7c35c12d1b6fd9e3fb653d1adff004f67

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
 
+### Hooks for signature validator
+
+Signature validation is a critical part of smart contract accounts. A flawed implementation of validation logic can lead to approving a transaction with an invalid signature. Signature validator hooks can be used to add safety net to the signature validation process and hold the power to block a signature validator from approving unauthorized actions. Hooks can be enabled for all validators befor ...
```
Is this hook called for the default signature validation?

akshay-ap replied:
yes

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+     * @notice A view function that the Manager will call when an account has enabled this contract as a function handler in the Manager for function isSignatureValid(bytes32,bytes)
+     * @param account Address of the account whose signature validator is to be used
+     * @param sender Address requesting signature validation
+     * @param data Calldata containing the 4 bytes function selector, 4 bytes signature selector, and one of the below:
```
Isn'T the signature selector only present in the 712 flow? Or are there different selectors?

akshay-ap replied:
There is only one selector for 712 for now.

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+     * @param domain bytes32 containing the domain for which Signature Validator contract should be used
+     * @param signatureValidatorContract Address of the Signature Validator Contract implementing ISafeProtocolSignatureValidator interface
+     */
+    function setSignatureValidator(bytes32 domain, address signatureValidatorContract) external;
```
this is the `domainSeparator` not the `domain` right?

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+     *              OR
+     *              encoded bytes containing bytes32 messageHash and bytes signatures
+     */
+    function handle(
```
As mentioned above, I would remove this method from the interface definition (and from this PR)

rmeissner commented on modules/README.md:
```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+### Security considerations
+
+A malicious actor can change the routing of signature validation flow by changing the 4 bytes of signature selector.
+Under such cases the signature validation should fail and revert the transaction. To prove that signature validation fails due to changed routing, formal verification can be used. This would be tackled in future.
```
Please create an issue for this.

akshay-ap replied:
https://github.com/safe-global/safe-core-protocol-specs/issues/47

rmeissner commented on modules/README.md:
```diff
@@ -232,7 +232,8 @@ The encoded data to be received by the signature validator manager as `bytes` pa
 function encodeData(bytes4 selector, bytes32 domain, bytes32 hashStruct, bytes calldata signatures) public pure returns (bytes memory) {
         bytes memory data = abi.encode(domain, hashStruct, signatures);
-        return abi.encodePacked(selector, data);
+        bytes4 selector = "safeSignature(bytes32,bytes32,bytes)";
```
In this case we do not need the `selector` as a parameter, right?

rmeissner reviewed (approved):
Minor comment about the `selector` param in `encodeData`
